### ISSUE TITLE
nix build should show more than the last 10 lines of output

### ISSUE THREAD

**Issue by chris-martin** [NixOS/nix#2420]
The Nix 2 `nix build` command truncates the log output for failed builds to 10 lines, which for most purposes (in my experience so far, at least) isn't enough to get any meaningful information. Should the `build` command have a flag to ask it nicely to print the entire output?

**Comment by chessai** [comment]
This has bitten me many times as well, I keep forgetting to open an issue about it. However, I don't think the flag should <i>enable</i> full logging, but rather it should <i>disable</i> it, full error messages seem like a more sensible default. When encountering an error, I'd rather see the whole thing by default, than not (and have to call log or something). Either way a flag would be useful

**Comment by ledettwy** [comment]
IMO nix build should behave just like nix-build except maintaining the progress bar at the bottom... that would solve this problem too

**Comment by copumpkin** [comment]
It seems like we could expand the current behavior to just show the full log on failure, right? I like not seeing noisy logs if all is well, but since Nix is storing the build logs, it might as well show them if the build fails.

**Comment by r-k-b** [comment]
This issue was one of the top results while searching for how to display the full log.
For posterity: you can see the log by copying the store path mentioned in `nix build`s output to the `nix store --read-log` command.

```
$ nix build .#dist-files
warning: Git tree '/home/rkb/projects/tulars' is dirty
building '/nix/store/f81k0pc8c1jii699g538hdz74cvrc8z1-dist-files.drv'...
error: --- Error -------------------------------------------------------------------------------------------------------------------------------------------------- nix
builder for '/nix/store/f81k0pc8c1jii699g538hdz74cvrc8z1-dist-files.drv' failed with exit code 2; last 10 log lines:
  npm ERR!   path: '/homeless-shelter'
  <...some other mangled log lines...>
```

```
$ nix-store --read-log /nix/store/f81k0pc8c1jii699g538hdz74cvrc8z1-dist-files.drv
```

**Comment by Kha** [comment]
Or, slightly shorter:
```
$ nix log /nix/store/f81k0pc8c1jii699g538hdz74cvrc8z1-dist-files.drv
```
Or, much shorter (but doesn't work if it's a dependency that failed):
```
$ nix log .#dist-files
```
And just for completeness, to show the full log during build whether it ultimately fails or not:
```
$ nix build -L .#dist-files
```

**Comment by flokli** [comment]
Still relevant.

**Comment by YellowOnion** [comment]
Anything is better than having to copy-paste that long url, it's not very user friendly.

**Comment by patricksjackson** [comment]
@YellowOnion if you want to see more lines in the failed output, just set the nix option `log-lines` to something larger.

This discussion is mostly a debate over whether `10` is a reasonable default for `log-lines` - https://nixos.org/manual/nix/stable/command-ref/conf-file.html?highlight=log-lines#conf-log-lines

**Comment by YellowOnion** [comment]
@patricksjackson Yeah I'm just saying, as a "I support raising the limit" sort of thing.

It's really bad work flow and slows down productivity quite a bit, and instead of just showing the entire log, like other compilation apps, it gives you this annoying file URL you have to copy-paste, so it's doubly worse, (the URL is nice if you have a compilation that blows out your terminal buffer tho).

Raising the limit and adding a `nix log --last` will probably go a long way to help.

**Comment by zopsicle** [comment]
> There is also `--option log-lines n`. If you set `n` to a high enough value, 1000000, that ought to give print all the logs as things fail.

I tried this but it does not work.

```console
$ nix-shell tools/nix/shell.nix --log-format bar --option log-lines 1000000 --pure --run true
these 2 derivations will be built:
  /nix/store/5b6y1zlz8cxyhydn4n0jyhivqywzx6sk-imagemagick-windows.drv
  /nix/store/36chvcz1gs8ja98cb0yvgn4b6353h80a-setup.drv
building '/nix/store/5b6y1zlz8cxyhydn4n0jyhivqywzx6sk-imagemagick-windows.drv'...
error: builder for '/nix/store/5b6y1zlz8cxyhydn4n0jyhivqywzx6sk-imagemagick-windows.drv' failed with exit code 1;
       last 10 log lines:
       >   -s [ --silent ]               Output only error/warning information
       >   --no-warn-unused              Don't warn on unused .bin files
       >   -c [ --color ] [=arg(=1)]     Enable/disable color output
       >   -p [ --progress ] [=arg(=1)]  Enable/disable the progress bar
       >
       > Extracts installers created by Inno Setup 1.2.10 to 6.0.5
       >
       > innoextract 1.9 (C) 2011-2020 Daniel Scharrer <daniel@constexpr.org>
       > This is free software with absolutely no warranty.
       > Running phase: installPhase
       For full logs, run 'nix-store -l /nix/store/5b6y1zlz8cxyhydn4n0jyhivqywzx6sk-imagemagick-windows.drv'.
error: 1 dependencies of derivation '/nix/store/36chvcz1gs8ja98cb0yvgn4b6353h80a-setup.drv' failed to build
```
